File: src/main/java/work/xeltica/craft/core/stores/PlayerStore.java
package work.xeltica.craft.core.stores;

import net.kyori.adventure.bossbar.BossBar;
import net.kyori.adventure.bossbar.BossBar.Color;
import net.kyori.adventure.bossbar.BossBar.Overlay;
import net.kyori.adventure.text.Component;
import org.bukkit.Bukkit;
import org.bukkit.FireworkEffect;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.entity.Player;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.FireworkMeta;
import org.bukkit.scheduler.BukkitTask;
import work.xeltica.craft.core.XCorePlugin;
import work.xeltica.craft.core.events.StaffJoinEvent;
import work.xeltica.craft.core.events.StaffLeaveEvent;
import work.xeltica.craft.core.models.PlayerDataKey;
import work.xeltica.craft.core.models.PlayerRecord;
import work.xeltica.craft.core.utils.Config;
import work.xeltica.craft.core.utils.Ticks;

import java.io.IOException;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.UUID;

/**
 * プレイヤー固有データの管理を行い、読み書きを行うPlayer Store APIを提供します。
 * @author Xeltica
 */
public class PlayerStore {
    public PlayerStore() {
        PlayerStore.instance = this;
        PlayerStore.liveBarMap = new HashMap<>();
        playerStores = new Config("playerStores");

        Bukkit.getScheduler().runTaskTimer(XCorePlugin.getInstance(), this::saveTask, 0, Ticks.from(10));
    }

    public static PlayerStore getInstance() {
        return PlayerStore.instance;
    }

    public static Map<UUID, BossBar> getliveBarMap() {
        return PlayerStore.liveBarMap;
    }

    public PlayerRecord open(OfflinePlayer p) {
        return open(p.getUniqueId());
    }

    public PlayerRecord open(UUID id) {
        var section = playerStores.getConf().getConfigurationSection(id.toString());
        if (section == null) {
            section = playerStores.getConf().createSection(id.toString());
        }
        return new PlayerRecord(playerStores, section, id);
    }

    public List<PlayerRecord> openAll() {
        return playerStores
                .getConf()
                .getKeys(false)
                .stream()
                .map(k -> open(UUID.fromString(k)))
                .toList();
    }

    public void save() throws IOException {
        playerStores.save();
    }

    public boolean isCitizen(Player p) {
        return p.hasPermission("otanoshimi.citizen");
    }

    public void setLiveMode(Player player, boolean isLive) {
        if (isLive == isLiveMode(player)) return;
        if (isLive) {
            final var name = String.format("%s が配信中", player.getName());
            final var bar = BossBar.bossBar(Component.text(name), BossBar.MAX_PROGRESS, Color.RED, Overlay.PROGRESS);

            liveBarMap.put(player.getUniqueId(), bar);
            BossBarStore.getInstance().add(bar);
        } else {
            final var bar = liveBarMap.get(player.getUniqueId());

            liveBarMap.remove(player.getUniqueId());
            BossBarStore.getInstance().remove(bar);
        }
    }

    public boolean isLiveMode(Player p) {
        return liveBarMap.containsKey(p.getUniqueId());
    }

    public ItemStack getRandomFireworkByUUID(UUID id, int amount) {
        final var random = new Random(id.hashCode());
        final var item = new ItemStack(Material.FIREWORK_ROCKET, amount);
        item.editMeta(meta -> {
            final var firework = (FireworkMeta) meta;
            final var effect = FireworkEffect.builder()
                    .trail(random.nextBoolean())
                    .flicker(random.nextBoolean())
                    .with(FireworkEffect.Type.values()[random.nextInt(5)])
                    .withColor(colors[random.nextInt(colors.length)])
                    .build();
            firework.addEffect(effect);
            firework.setPower(1);
        });
        return item;
    }

    public boolean isOnline(Player player) {
        return isOnline(player.getUniqueId());
    }

    public boolean isOnline(UUID id) {
        return Bukkit.getOnlinePlayers().stream().anyMatch(p -> p.getUniqueId().equals(id));
    }

    private void saveTask() {
        if (!isChanged()) return;
        try {
            save();
            setChanged(false);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static PlayerStore instance;
    private final Config playerStores;
    private static Map<UUID, BossBar> liveBarMap;

    private boolean changed;

    private static final org.bukkit.Color[] colors = {
            // from XelticaUI
            org.bukkit.Color.fromRGB(0xe23731), // red
            org.bukkit.Color.fromRGB(0xeb6101), // vermilion
            org.bukkit.Color.fromRGB(0xf08300), // orange
            org.bukkit.Color.fromRGB(0xe9be00), // yellow
            org.bukkit.Color.fromRGB(0xb8d200), // lime
            org.bukkit.Color.fromRGB(0x3eb370), // green
            org.bukkit.Color.fromRGB(0x20c0a0), // teal
            org.bukkit.Color.fromRGB(0x43fcf3), // cyan
            org.bukkit.Color.fromRGB(0x00b7ff), // skyblue
            org.bukkit.Color.fromRGB(0x2571ff), // blue
            org.bukkit.Color.fromRGB(0xff55a1), // magenta
            org.bukkit.Color.fromRGB(0xff5c84), // pink
    };

    public boolean isChanged() {
        return this.changed;
    }

    public void setChanged(boolean changed) {
        this.changed = changed;
    }
}

File: src/main/java/work/xeltica/craft/core/stores/PlayerStore.kt
package work.xeltica.craft.core.stores

import net.kyori.adventure.bossbar.BossBar
import net.kyori.adventure.text.Component
import org.bukkit.*
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.FireworkMeta
import org.bukkit.inventory.meta.ItemMeta
import work.xeltica.craft.core.api.Ticks.from
import work.xeltica.craft.core.modules.BossBarModule.add
import work.xeltica.craft.core.modules.BossBarModule.remove
import work.xeltica.craft.core.models.PlayerRecord
import kotlin.Throws
import java.io.IOException
import work.xeltica.craft.core.XCorePlugin
import work.xeltica.craft.core.api.Config
import java.lang.Runnable
import java.util.*

/**
 * プレイヤー固有データの管理を行い、読み書きを行うPlayer Store APIを提供します。
 * @author Xeltica
 */
class PlayerStore {
    fun open(p: OfflinePlayer): PlayerRecord {
        return open(p.uniqueId)
    }

    fun open(id: UUID): PlayerRecord {
        var section = playerStores.conf.getConfigurationSection(id.toString())
        if (section == null) {
            section = playerStores.conf.createSection(id.toString())
        }
        return PlayerRecord(playerStores, section, id)
    }

    fun openAll(): List<PlayerRecord> {
        return playerStores
            .conf
            .getKeys(false)
            .stream()
            .map { k: String? -> open(UUID.fromString(k)) }
            .toList()
    }

    @Throws(IOException::class)
    fun save() {
        playerStores.save()
    }

    fun isCitizen(p: Player): Boolean {
        return p.hasPermission("otanoshimi.citizen")
    }

    fun setLiveMode(player: Player, isLive: Boolean) {
        if (isLive == isLiveMode(player)) return
        if (isLive) {
            val name = String.format("%s が配信中", player.name)
            val bar =
                BossBar.bossBar(Component.text(name), BossBar.MAX_PROGRESS, BossBar.Color.RED, BossBar.Overlay.PROGRESS)
            liveBarMap!![player.uniqueId] = bar
            add(bar)
        } else {
            val bar = liveBarMap!![player.uniqueId]
            liveBarMap!!.remove(player.uniqueId)
            remove(bar!!)
        }
    }

    fun isLiveMode(p: Player): Boolean {
        return liveBarMap!!.containsKey(p.uniqueId)
    }

    fun getRandomFireworkByUUID(id: UUID, amount: Int): ItemStack {
        val random = Random(id.hashCode().toLong())
        val item = ItemStack(Material.FIREWORK_ROCKET, amount)
        item.editMeta { meta: ItemMeta ->
            val firework = meta as FireworkMeta
            val effect = FireworkEffect.builder()
                .trail(random.nextBoolean())
                .flicker(random.nextBoolean())
                .with(FireworkEffect.Type.values()[random.nextInt(5)])
                .withColor(colors[random.nextInt(colors.size)])
                .build()
            firework.addEffect(effect)
            firework.power = 1
        }
        return item
    }

    fun isOnline(player: Player): Boolean {
        return isOnline(player.uniqueId)
    }

    fun isOnline(id: UUID): Boolean {
        return Bukkit.getOnlinePlayers().stream().anyMatch { p: Player -> p.uniqueId == id }
    }

    private fun saveTask() {
        if (!isChanged) return
        try {
            save()
            isChanged = false
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private val playerStores: Config
    var isChanged = false

    init {
        instance = this
        liveBarMap = HashMap()
        playerStores = Config("playerStores")
        Bukkit.getScheduler().runTaskTimer(XCorePlugin.instance, Runnable { saveTask() }, 0, from(10.0).toLong())
    }

    companion object {
        @JvmStatic
        fun getliveBarMap(): Map<UUID, BossBar>? {
            return liveBarMap
        }

        @JvmStatic
        lateinit var instance: PlayerStore
            private set
        private var liveBarMap: MutableMap<UUID, BossBar>? = null
        private val colors = arrayOf( // from XelticaUI
            Color.fromRGB(0xe23731),  // red
            Color.fromRGB(0xeb6101),  // vermilion
            Color.fromRGB(0xf08300),  // orange
            Color.fromRGB(0xe9be00),  // yellow
            Color.fromRGB(0xb8d200),  // lime
            Color.fromRGB(0x3eb370),  // green
            Color.fromRGB(0x20c0a0),  // teal
            Color.fromRGB(0x43fcf3),  // cyan
            Color.fromRGB(0x00b7ff),  // skyblue
            Color.fromRGB(0x2571ff),  // blue
            Color.fromRGB(0xff55a1),  // magenta
            Color.fromRGB(0xff5c84)
        )
    }
}